File: UtilitiesVB/Common Outlook Functions/ExplorerActions.cs
using Microsoft.Office.Interop.Outlook;
using System;
using System.Collections.Generic;
using System.Collections;
using System.Diagnostics;
using System.Linq;


namespace UtilitiesVB
{
    public static class ExplorerActions
    {
        public static object GetCurrentItem(Application OlApp)
        {
            if (OlApp == null) { throw new ArgumentNullException(nameof(OlApp)); }
            else if (OlApp.ActiveWindow() is Explorer) { return Readable(OlApp.ActiveExplorer().Selection[0]); }
            else if (OlApp.ActiveWindow() is Inspector) { return Readable(OlApp.ActiveInspector().CurrentItem); }
            else { return null; }
        }

        internal static object Readable(object item)
        {
            if ((item is MailItem) && ((MailItem)item).IsMailUnReadable())
            { return null; }
            return item;
        }

    }
}

File: UtilitiesVB/Common Outlook Functions/ExplorerActions.vb
Imports Microsoft.Office.Interop
Imports Microsoft.Office.Interop.Outlook

Public Module ExplorerActions
    Public Function GetCurrentItem(OlApp As Outlook.Application) As Object

        Dim oMail As MailItem
        Dim obj As Object


        On Error Resume Next
        Select Case TypeName(OlApp.ActiveWindow)
            Case "Explorer"
                obj = OlApp.ActiveExplorer.Selection.Item(1)
            Case "Inspector"
                obj = OlApp.ActiveInspector.CurrentItem
            Case Else
                obj = Nothing
        End Select

        If TypeOf obj Is Outlook.MailItem Then
            oMail = obj
            If IsMailUnReadable(oMail) Then
                obj = Nothing
            End If
        End If

        Return obj

    End Function

End Module
